### ISSUE TITLE
More trackingID: L2, L3,...

### ISSUE THREAD

**Issue by nlewo** [skydive-project/skydive#196]
We currently have a trackingID that hashes all layers from the L2 layer. This is a L2TrackingID. It would be great to also have L3. The L3TrackingID would hash all layers from layer L3.

We could also define them as following:
```
L2TrackingID = hash(L2_Layer, L3TrackingID)
  where L2_Layer = hash(mac_src, mac_dst)
L3TrackingID = hash(L3_Layer, L7TrackingID)
  where L3_Layer = hash(ip_src, ip_dst, more-info)
```

So, the L2TrackingID is more precise than the L3TrackingID.

# Motivation

To forward a packet, Contrail can remove the Ethernet header when it is useless. In some cases, when the packet is encapsulated in the MPLS tunnel, the Ethernet header is removed. So, trackingIDs of the packet captured on the tap and on the tunnel interface are not the same. We are then not able to follow However, the L3TrackingID would be the same on both interface.

As a bonus, we could also define a L7TrackingID. For instance, it could only hash a subset of the HTTP header. This would provide the ability to follow packets across loadbalancer for instance. But this has to be addressed by another PR!

**Comment by safchain** [comment]
I think that is a pretty good idea, I think we could start just with l2, l3/l4 to see the inpact in term of API

**Comment by nlewo** [comment]
Ok, so let's start ;)

**Comment by nplanel** [comment]
Technically we can fix this issue (L2 missing due to MPLS strip ethernet) by splitting all flows on L2
So we will have the same behavior than tunnels

The query could be a little bit more complex (or simplify some use case) as multiple L3/L4 flows could have the same signature (TrackingID) but could be differentiated, thanks to parentUUID.

**Comment by safchain** [comment]
FYI for the parent UUID filtering is was the intent of this patch

https://softwarefactory-project.io/r/#/c/5852/

**Comment by nlewo** [comment]
@nplanel yep, this could solve the issue. But this will also increase the number of flows. Moreover, I think the current flow creation is natural:
```
1 packet -> 1 flow
1 packet encapsulated in 1 packet -> 2 flows
...
```

If we split them at the l2 layer, I think it would become a little bit harder to understand how flows are created and what they represent.

**Comment by nplanel** [comment]
The split logic is already on the Network layer so it's make sens to keep this rule generic.
Multiple flows will not be a big deal as it would be a updated in the DB.

I think it would clarify the understanding as the all the flows informations (L7 proto, metrics ...) are on the top layers, and If you need more info chained information will be pointed by parentUUID flow

**Comment by safchain** [comment]
@nplanel not only in the DB we will have multiple flows in the flow tables.

**Comment by nplanel** [comment]
@safchain yes sure it was just as an example as DB is more likely a "bottleneck" compare to flow tables

**Comment by nlewo** [comment]
Closed by 843b333b52d085abd8b51274fc2f7e3a23aa85ad
